src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -662,11 +811,53 @@ pub fn is_useful<'p, 'a: 'p, 'tcx: 'a>(cx: &mut MatchCheckCtxt<'a, 'tcx>,
             pat_constructors(cx, row[0], pcx).unwrap_or(vec![])
         }).collect();
         debug!("used_ctors = {:#?}", used_ctors);
+        // `all_ctors` are all the constructors for the given type, which
+        // should all be represented (or caught with the wild pattern `_`).
         let all_ctors = all_constructors(cx, pcx);
         debug!("all_ctors = {:#?}", all_ctors);
-        let missing_ctors: Vec<Constructor> = all_ctors.iter().filter(|c| {
-            !used_ctors.contains(*c)
-        }).cloned().collect();
+
+        // The only constructor patterns for which it is valid to
+        // treat the values as constructors are ranges (see
+        // `all_constructors` for details).
+        let exhaustive_integer_patterns = cx.tcx.features().exhaustive_integer_patterns;
+        let consider_value_constructors = exhaustive_integer_patterns
+            && all_ctors.iter().all(|ctor| match ctor {
+                ConstantRange(..) => true,
+                _ => false,
+            });
+
+        // `missing_ctors` are those that should have appeared
+        // as patterns in the `match` expression, but did not.
+        let mut missing_ctors = vec![];
+        for req_ctor in &all_ctors {
+            if consider_value_constructors {
+                let mut refined_ctors = vec![req_ctor.clone()];
+                for used_ctor in &used_ctors {
+                    // Refine the required constructors for the type by subtracting
+                    // the range defined by the current constructor pattern.
+                    refined_ctors = match IntRange::from_ctor(cx.tcx, used_ctor) {
+                        Some(interval) => interval.subtract_from(cx.tcx, refined_ctors),
+                        None => refined_ctors,
+                    };
+                    // If the constructor patterns that have been considered so far
+                    // already cover the entire range of values, then we the
+                    // constructor is not missing, and we can move on to the next one.
+                    if refined_ctors.is_empty() {
+                        break;
+                    }
+                }
+                // If a constructor has not been matched, then it is missing.
+                // We add `refined_ctors` instead of `req_ctor`, because then we can
+                // provide more detailed error information about precisely which
+                // ranges have been omitted.
+                missing_ctors.extend(refined_ctors);
+            } else {
+                // A constructor is missing if it never appears in a `match` arm.
+                if !used_ctors.iter().any(|used_ctor| used_ctor == req_ctor) {
+                    missing_ctors.push(req_ctor.clone());
+                }
```

Comment:
I think these codepaths could be unified - that is, in the other path, if `exhaustive_integer_patterns` is `false` or `IntRange::from_ctor` returns `None`, *also* try `==`.
I doubt `consider_value_constructors` is needed at all, given those changes.